snapatac2-contrib/cemba_metric/cemba_metrics.py

```diff
@@ -0,0 +1,203 @@
+
+Here we include three metrics:
+1. PAC, proportion of ambiguous clustering
+   - it's \in (0,1)
+   - the lower, the better
+   - (\sum (M_{ij} < 0.95) - \sum (M_{ij} < 0.05) ) / n_cell ^2
+2. Dispersion coefficient (disp)
+   - it's \in (0,1)
+   - the higher, the better
+   - \sum 4 * (M_{ij} - 0.5)^2 / n_cell ^2.
+
+3. Cumulative distribution function (cdf) curve 
+  - draw the cdf based on the cdf of M.
+  - since M_{ij} should be around 0 or 1 ideally. So the cdf curve would
+    be have a relative flat region in the middle,
+    and sharp increase around 0.0 and 1.0.
+
+During analysis, we use PAC and disp to choose the resolution, which
+will be the values of cemba_metrics.
+"""
+
+import os
+import scipy
+import random
+from typing import Tuple, Union
+from scipy.sparse import csr_matrix
+import numpy as np
+import itertools
+
+
+def cal_connectivity(P):
+    """calculate connectivity matrix"""
+    connectivity_mat = np.zeros((len(P), len(P)), dtype = bool)
+    classN = max(P)
+    ## TODO: accelerate this
+    for cls in range(int(classN + 1)):
+        xidx = [i for i, x in enumerate(P) if x == cls]
+        iterables = [xidx, xidx]
+        for t in itertools.product(*iterables):
+            connectivity_mat[t[0], t[1]] = True
+    """connectivity_mat = csr_matrix(connectivity_mat)"""
+    return connectivity_mat
+
+
+def reorder(C):
+    """
+    Reorder consensus matrix.
+
+    :param C: Consensus matrix.
+    :type C: `numpy.ndarray`
+    """
+    from scipy.spatial.distance import squareform
+    from scipy.cluster.hierarchy import linkage, leaves_list
+    Y = 1 - C
+    Z = linkage(squareform(Y), method="average")
+    ivl = leaves_list(Z)
+    ivl = ivl[::-1]
+    return C[:, ivl][ivl, :]
+
+def plot_CDF(prefix, C, u1, u2, num_bins=100):
+    import matplotlib.pyplot as plt
+    counts, bin_edges = np.histogram(C, bins=num_bins, density=True)
+    cdf = np.cumsum(counts)
+    fig = plt.figure(dpi=100)
+    plt.plot(bin_edges[1:], cdf)
+    plt.xlabel("Consensus index value")
+    plt.ylabel("CDF")
+    plt.axvline(x=u1, color="grey", linestyle="--")
+    plt.axvline(x=u2, color="grey", linestyle="--")
+    fileN = [prefix, "cdf", "png"]
+    fileN = ".".join(fileN)
+    fig.savefig(fileN)
+    outBinEdges = ".".join([prefix, "cdf.txt"])
+    with open(outBinEdges, "w") as fo:
+        fo.write("\t".join(str(i) for i in cdf) + "\n")
+
+
+def cal_dispersion(C):
+    """calculate dispersion coefficient"""
+    n = C.shape[1]
+    corr_disp = np.sum(
+        4 * np.square(C - 0.5), dtype = np.float64) / (np.square(n))
+    return corr_disp
+
+
+def cal_PAC(C, u1, u2):
+    """calculate PAC (proportion of ambiguous clustering)"""
+    n = C.shape[0] ** 2
+    PAC = ((C<u2).sum() - (C<u1).sum()) / n
+    return PAC
+
+def cal_stab(x) -> np.ndarray:
+    """calculate stability for every cell"""
+    s = np.sum(abs(x - 0.5)) / (0.5 * x.shape[0])
+    return s
+
+
+def cemba_metric(
```

Comment:
What are the meanings of output (disp, PAC)? How am I supposed to use these values?

Reply:
I introduce the meaning of disc and PAC at the top of the script. You can treat it as independent measurements.